src/board_builder/board_builder.py

```diff
@@ -0,0 +1,258 @@
+
+        ### MATRIX INIT ###
+        self._matrix_index = 0
+        self._relative_pose_matrix = [[None]]
+        # TODO: Will need to be calculated rather than hard coded
+        self.MARKER_MIDDLE_TO_EDGE_IN_PIXELS = 20
+        self.local_corners = np.array([
+            [-self.MARKER_MIDDLE_TO_EDGE_IN_PIXELS, self.MARKER_MIDDLE_TO_EDGE_IN_PIXELS, 0, 1],  # Top-left
+            [self.MARKER_MIDDLE_TO_EDGE_IN_PIXELS, self.MARKER_MIDDLE_TO_EDGE_IN_PIXELS, 0, 1],  # Top-right
+            [self.MARKER_MIDDLE_TO_EDGE_IN_PIXELS, -self.MARKER_MIDDLE_TO_EDGE_IN_PIXELS, 0, 1],  # Bottom-right
+            [-self.MARKER_MIDDLE_TO_EDGE_IN_PIXELS, -self.MARKER_MIDDLE_TO_EDGE_IN_PIXELS, 0, 1]  # Bottom-left
+        ])
+
+    def _calculate_corners_location(self, T_matrix, local_corners):
+        """ Given a matrix transformation, find the location of the four corners """
+        corners_reference = np.zeros((4, 4))
+        for i in range(4):
+            corners_reference[i] = T_matrix @ local_corners[i]
+
+        corners_reference = corners_reference[:, :3]
+        return corners_reference
+
+    def _calculate_relative_transform(self, T1, T2):
+        """ Given transform T1 from reference to marker 1, and transfrom T2 from reference to marker 2, calculate the
+        transform from T1 to T2"""
+        T1_inv = np.linalg.inv(T1)
+        T_relative = np.dot(T1_inv, T2)
+        return T_relative
+
+    def _estimate_reference_to_occluded(self, T1, T2):
+        """ Given transform T1 from reference to visible marker, and transform T2 from visible marker to occluded
+        marker, calculate the transform from the reference to the occluded marker"""
+        T_reference_to_occluded = np.dot(T1, T2)
+        return T_reference_to_occluded
+
+    def _expand_matrix(self):
+        """ Adds one row and one column to the matrix and initializes them to None """
+        size = len(self._relative_pose_matrix) + 1
+        new_matrix = [[None for _ in range(size)] for _ in range(size)]
+        for i in range(size - 1):
+            for j in range(size - 1):
+                new_matrix[i][j] = self._relative_pose_matrix[i][j]
+        self._relative_pose_matrix = new_matrix
+
+    def _filter_data(self, data):
+        """ Filters the detector data to only keep the markers snapshots that appear in more than one detector """
+        label_counts = defaultdict(int)
+        for snapshots in data.values():
+            for snapshot in snapshots:
+                label_counts[snapshot.label] += 1
+
+        filtered_data = {}
+        for key, snapshots in data.items():
+            filtered_snapshots = [
+                snapshot for snapshot in snapshots
+                if label_counts[snapshot.label] >= 2
+            ]
+            if filtered_snapshots:
+                filtered_data[key] = filtered_snapshots
+        return filtered_data
+
+    def _find_matrix_input_index(self, pose_uuid, other_pose_uuid):
```

Comment:
> Ok, so just to make sure I understand, this is just mapping from the uuid to the column/row index, am I right?

Yes, that's correct.



> I wonder if it's worth making up a term for this column/row index. Or just refer to it consistently as a `matrix_index` whenever we use this concept.

I think for now it's fine because there isn't any ambiguity as what it is doing and 'matrix index' makes it clear we are searching for a pose in the matrix. I am open to variable name suggestions though.

Reply:
`matrix_index` sounds fine to me.